Eth ABI: Rework for Android usage (JNI, no templates).

catenocrypt opened the pull request:
## Description

Eth ABI support has been extended to be usable from Android (and iOS) project.
Fixes #573 .
- Eth ABI encode has been reworked, with no templated implementation, as template typing is lost through C/JNI iface
- Decode support functionality added
- C and JNI interface added

## Testing instructions

- Encode/decode unit tests added.
- Added test through the native interface.
- Java side tests -- TODO

## Types of changes

* Refactor of Eth ABI implementaiton
* Extension of Eth ABI implementation

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.

catenocrypt commented:
Current status:
- C++ project compiles
- C++ unit tests are OK
- codegen is OK
- wallet-core/android part (incl. generated JNI code) compiles

TODO:
- add tests on android side
- validate by integrating into android app, use it from there

hewigovens commented:
FYI, I disabled kotlin rule: StringLiteralDuplication

catenocrypt commented:
> FYI, I disabled kotlin rule: StringLiteralDuplication
I already fixed this (not pushed), and it does make it more readable, so I'm fine with the rule :)

catenocrypt commented:
Ready for review, at last (squash into a monster commit)

hewigovens commented on android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/ethereum/TestEthereumAbiEncoder.kt:
```diff
@@ -0,0 +1,65 @@
+        val encoded = EthereumAbiEncoder.encode(function)
+        assertEquals("0xa5643bf2" +
+            "0000000000000000000000000000000000000000000000000000000000000060" +
+            str32_1 +
```
Variables make it hard to read the test, since we disable the kotlin rule, can we use plain strings?

hewigovens commented on include/TrustWalletCore/TWEthereumAbiEncoder.h:
```diff
@@ -0,0 +1,42 @@
+TW_EXPORT_STATIC_METHOD
+struct TWEthereumAbiFunction *_Nullable TWEthereumAbiEncoderBuildFunction(TWString *_Nonnull name);
+
+/// Deletes a function object created with a 'TWEthereumAbiEncoderCreateWithString' method.
```
```suggestion
/// Deletes a function object created with a 'TWEthereumAbiEncoderBuildFunction' method.
```

hewigovens commented on src/Ethereum/ABI/Array.cpp:
```diff
@@ -0,0 +1,84 @@
+using namespace TW::Ethereum::ABI;
+
+int ParamArray::addParam(const std::shared_ptr<ParamBase>& param) {
+    if (param == nullptr) return -1;
```
we prefer always have `{}`, `{ return -1; }`

hewigovens commented on src/Ethereum/ABI/Array.cpp:
```diff
@@ -0,0 +1,84 @@
+
+void ParamArray::encode(Data& data) const {
+    size_t n = _params.getCount();
+    TW::Ethereum::ABI::encode(uint256_t(n), data);
```
```suggestion
    ABI::encode(uint256_t(n), data);
```

hewigovens commented on src/Ethereum/ABI/Array.cpp:
```diff
@@ -0,0 +1,84 @@
+    for (auto i = 0; i < n; ++i) {
+        auto p = _params.getParamUnsafe(i);
+        if (p->isDynamic()) {
+            TW::Ethereum::ABI::encode(uint256_t(headSize + dynamicOffset), data);
```
```suggestion
            ABI::encode(uint256_t(headSize + dynamicOffset), data);
```

hewigovens commented on src/Ethereum/ABI/Array.cpp:
```diff
@@ -0,0 +1,84 @@
+}
+
+void ParamArray::addParams(const std::vector<std::shared_ptr<ParamBase>>& params) {
+    for (auto p: params) addParam(p);
```
```suggestion
    for (auto p: params) { addParam(p); }
```

hewigovens commented on src/Ethereum/ABI/Array.cpp:
```diff
@@ -0,0 +1,84 @@
+    }
+    // padding
+    offset_inout = origOffset + Util::paddedTo32(offset_inout - origOffset);
+    // TODO handle dynamic types
```
create issues to track?

hewigovens commented on src/Ethereum/ABI/Function.h:
```diff
@@ -1,73 +1,82 @@
-// Copyright © 2017-2019 Trust Wallet.
+// Copyright © 201§-2019 Trust Wallet.
```
```suggestion
// Copyright © 2017-2019 Trust Wallet.
```

hewigovens commented on src/Ethereum/ABI/ParamNumber.cpp:
```diff
@@ -0,0 +1,63 @@
+
+bool ParamIntN::decodeNumber(const Data& encoded, int256_t& decoded, size_t& offset_inout) {
+    uint256_t valU;
+    auto res = TW::Ethereum::ABI::decode(encoded, valU, offset_inout);
```
```suggestion
    auto res = ABI::decode(encoded, valU, offset_inout);
```

hewigovens commented on tests/interface/TWEthereumAbiEncoderTests.cpp:
```diff
@@ -0,0 +1,233 @@
+    TWEthereumAbiEncoderDeleteFunction(func);
+}
+
+TEST(TWEthereumAbi, EncodeFuncMonster) {
```
👍

vikmeup commented on src/Ethereum/ABI/ParamNumber.cpp:
```diff
@@ -0,0 +1,63 @@
+
+void ParamIntN::init()
+{
+    assert(bits >= 8 && bits <= 256 && (bits % 8) == 0);
```
@hewigovens is it safe to use assert in production?

hewigovens replied:
assert won't work in release, the app needs to prepare the data correctly

catenocrypt replied:
Assert have to be used with care.  They are optimized out from release build.  I use it only in fatal, programming error, 'never can happen' cases.  Sometimes I use them in incorrect input case, but also handling it nicely, like this:
    assert(param != nullptr);
    if (param == nullptr) { return -1; }
This helps catching problems in debug/tests.

catenocrypt commented:
Incorporated review comments (thanks, @hewigovens )
Decoding part was added to the Kotlin test, now I have also added it to the Swift test (for decoding function output values).

hewigovens commented:
/azp run
